✨ Add Mermaid Support to Rich Text Field (Plate)

JackDevAU opened the pull request:
#4734 

> [!NOTE]
> TODO: Write a docs page explaining how to use mermaid in TinaCMS

- Adds mermaid support to Plate editor
	- Preview + edit mermaid diagrams inside a Rich Text Field
	- Saves as ```mermaid codeblock (as per the Standard)
	-  Light mode and dark mode to view mermaid diagram previews better
	- Internal - New type added to Plate - `mermaid`
	- Ability to hide `mermaid` via `toolbarOverride` 
	- Added a test for mermaid diagram
	- Shows errors 
- `<TinaMarkdown />` now has a `mermaid` property that passes the mermaid diagram as a `value` that can then be used to render however the end user wants. Example uses [mermaid npm package](https://www.npmjs.com/package/mermaid)
- Fixes Tooltip's not showing correctly.  

![new-mermaid](https://github.com/user-attachments/assets/9c5f35c9-09eb-4b5b-a200-67174093762a)
**Figure: Mermaid support inside Rich Text Field**

<img width="589" alt="image" src="https://github.com/user-attachments/assets/6f7ea26e-3b6f-426a-b399-9a8e515d840c">
**Figure: Mermaid Error**

ncn-ssw commented on packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  useEffect(() => {
+    //? Note: Adding this class on a <pre/> will remove its base bg styling
```
**nitpick**: Stray '?'
```suggestion
    // Note: Adding this class on a <pre/> will remove its base bg styling
```

JackDevAU replied:
Ah this is due to an extension that I use - https://marketplace.visualstudio.com/items?itemName=aaron-bond.better-comments 
Makes the comment blue when adding a ?

ncn-ssw commented:
**issue**: By default, the component is being rendered in dark mode.

This looks very incongruent with the rest of the current administration interface.

![image](https://github.com/user-attachments/assets/eb8b0afb-d1b5-47d4-a265-f197575d7ae8)
**Figure: The dark box with difficult to see lines**

ncn-ssw commented:
**issue**: The graph is not appearing in the live preview, only its text form.

I expected to see the graph in the live preview when editing was complete on the graph itself.

![image](https://github.com/user-attachments/assets/622a0490-e1db-4cb9-98a5-af9a2d58e1de)
**Figure: The appearance of the block**

ncn-ssw commented:
**issue**: Saving a mermaid diagram fails with the error message "BlockElement: mermaid is not yet supported"

![image](https://github.com/user-attachments/assets/157ece56-a6ef-4b06-8ac7-bc7f3eb605f7)  
**Figure: The error dialog shown when the page "Save" button was clicked**

ncn-ssw commented:
**issue**: The initial form of the Mermaid diagram is a "Syntax Error" message - I expected either a graph or some form of guidance.

The example graph I used below is as follows:

```
graph TD;
    id1(Click the ✏️ edit button at top right of mermaid component) --> id2(Modify this graph);
    id2 --> id3(Click edit button again);
```

![image](https://github.com/user-attachments/assets/f349c22d-54b4-4edf-8f92-29670a0c4266)  
**Figure: The "Syntax Error" message from a newly inserted Mermaid graph at the arrow**

ncn-ssw commented:
**issue**: When there is a syntax error, there is no indication _what_ the error is or any method by which to retrieve the error.

I expected that there would be some sort of useful message.

![image](https://github.com/user-attachments/assets/9b9844c6-ac3a-4783-b139-371892f9c357)  
**Figure: Computer says 'no', doesn't elaborate**

ncn-ssw commented:
**issue**: Dark-mode is illegible.

I expected that the default color of text would become lighter, rather than remain the same as in light mode.

![image](https://github.com/user-attachments/assets/6dba8499-2ca7-474c-a471-b41517eb2a00)
**Figure: Syntax error message is barely visible**

JackDevAU commented:
Reviewed the feedback @ncn-ssw 🙏🙏

Added and changed the following, 

- Mermaid diagram's now have a default diagram 
- Made light mode the default 
- Buttons now have a nice blue background and show on hover/click (on mobile) 
- Buttons hide when not hovered
- mermaid diagram defaults to code editor when clicked - instead of showing the syntax warning
- Added a "preview" icon instead of just leaving the edit "pencil" when switching between code and mermaid preview 

![image](https://github.com/user-attachments/assets/99de00a5-2f4a-48c9-bfee-d1502972d76f)
**Figure: Mermaid default diagram**

![new-mermaid](https://github.com/user-attachments/assets/b1c8a035-dd35-4f81-8e03-e48ac536c0b2)
**Figure: New edit and light/dark mode buttons**

JackDevAU commented:
> issue: The graph is not appearing in the live preview, only its text form.
> I expected to see the graph in the live preview when editing was complete on the graph itself.
> @ncn-ssw 

This is intentional - we are only shipping mermaid support and rendering inside the TinaCMS preview panel. It is up to the end user on how they want to render the mermaid diagram. 

As per my discussion with Jeff we both feel this is the best way as there can be multiple different ways to render mermaid. I'll have what we are using here in the example to test mermaid diagrams documented in the tina.io docs (With code examples). This will also include some other links and methods that can be used to render mermaid diagrams. e.g. remark etc.

JackDevAU commented on examples/next-2024/content/posts/hello-world.mdx:
````diff
@@ -1,9 +1,40 @@
 
 Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s  ...
+
+```mermaid
+flowchart TD
+    id1(this is an example flow diagram)
+    --> id2(modify me to see changes!)
+    id2
+    --> id3(Click the top button to preview the changes)
+    --> id4(Learn about mermaid diagrams @ mermaid.js.org)
+```
````
```suggestion

```

we only need 1 diagram for testing!

JackDevAU commented on packages/@tinacms/mdx/src/next/stringify/acorn.ts:
```diff
@@ -200,6 +200,7 @@ export function stringifyProps(
         if (field.list) {
           throw new Error(`Rich-text list is not supported`)
         } else {
+          console.log('RICH-TEXT StringifyProps', value)
```
```suggestion

```

ncn-ssw reviewed (dismissed):
🦙 Approved - Excellent work

JackDevAU commented on packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx:
```diff
@@ -0,0 +1,128 @@
+  return null
+}
+
+const DEFAULT_MERMAID_CONFIG = `%% This won't render without implementing a rendering engine (e.g. mermaid on npm)
+flowchart TD
+    id1(this is an example flow diagram) 
+    --> id2(modify me to see changes!)
+    id2 
+    --> id3(Click the top button to preview the changes)
+    --> id4(Learn about mermaid diagrams @ mermaid.js.org)`
```
```suggestion
const DEFAULT_MERMAID_CONFIG = `
flowchart TD
    id1(this is an example flow diagram) 
    --> id2(modify me to see changes!)
    id2 
    --> id3(Click the top button to preview the changes)
    --> id4(Learn about mermaid diagrams @ mermaid.js.org)`
```

JackDevAU commented on .changeset/cold-buses-bow.md:
```diff
@@ -0,0 +1,8 @@
+---
+"tinacms": patch
+"@tinacms/mdx": patch
+"@tinacms/schema-tools": patch
```
```suggestion
"tinacms": patch
"@tinacms/cli": patch
"@tinacms/mdx": patch
"@tinacms/schema-tools": patch
```

ncn-ssw reviewed (dismissed):
🦙

ncn-ssw reviewed (approved):
🦙
